scripts/performance-helper.psm1

```diff
@@ -0,0 +1,410 @@
+            $RemoteMasked = $ToMaskRemoteAddress -band $Mask
+
+            if ($RemoteMasked -eq $LocalSubnet) {
+                $MatchedIPs += $LocalIpAddr.IPAddressToString
+            }
+        }
+    }
+
+    if ($MatchedIPs.Length -ne 1) {
+        Write-Error "Failed to parse local address matching remote"
+    }
+
+    return $MatchedIPs[0]
+}
+
+function Invoke-TestCommand {
+    param ($Session, $ScriptBlock, [Object[]]$ArgumentList = @(), [switch]$AsJob = $false)
+
+    if ($Local) {
+        if ($AsJob) {
+            return Start-Job -ScriptBlock $ScriptBlock -ArgumentList $ArgumentList
+        }
+        return Invoke-Command -ScriptBlock $ScriptBlock -ArgumentList $ArgumentList
+    } else {
+        if ($AsJob) {
+            return Invoke-Command -Session $Session -ScriptBlock $ScriptBlock -AsJob -ArgumentList $ArgumentList
+        }
+        return Invoke-Command -Session $Session -ScriptBlock $ScriptBlock -ArgumentList $ArgumentList
+    }
+}
+
+function Wait-ForRemoteReady {
+    param ($Job, $Matcher)
+    $StopWatch =  [system.diagnostics.stopwatch]::StartNew()
+    while ($StopWatch.ElapsedMilliseconds -lt 10000) {
+        $CurrentResults = Receive-Job -Job $Job -Keep
+        if (![string]::IsNullOrWhiteSpace($CurrentResults)) {
+            $DidMatch = $CurrentResults -match $Matcher
+            if ($DidMatch) {
+                return $true
+            }
+        }
+        Start-Sleep -Seconds 0.1 | Out-Null
+    }
+    return $false
+}
+
+function Wait-ForRemote {
+    param ($Job)
+    Wait-Job -Job $Job -Timeout 10 | Out-Null
+    $RetVal = Receive-Job -Job $Job
+    return $RetVal -join "`n"
+}
+
+function Copy-Artifacts {
+    [Diagnostics.CodeAnalysis.SuppressMessageAttribute('PSAvoidUsingEmptyCatchBlock', '')]
+    param ([string]$From, [string]$To)
+    try {
+        Invoke-TestCommand $Session -ScriptBlock {
+            param ($To)
+            Remove-Item -Path "$To/*" -Recurse -Force
+        } -ArgumentList $To
+    } catch {
+        # Ignore failure, which occurs when directory does not exist
+    }
+    # TODO Figure out how to filter this
+    Copy-Item -Path "$From\*" -Destination $To -ToSession $Session  -Recurse
+}
+
+class ArgumentsSpec {
+    [string]$All;
+    [string]$Loopback;
+    [string]$Remote;
+
+    [string]GetArguments() {
+        if ($script:Local) {
+            return "$($this.All) $($this.Loopback)"
+        } else {
+            return "$($this.All) $($this.Remote)"
+        }
+    }
+}
+
+class ExecutableSpec {
+    [string]$Platform;
+    [string[]]$Tls;
+    [string[]]$Arch;
+    [string]$Exe;
+    [ArgumentsSpec]$Arguments;
+}
+
+class TestDefinition {
+    [string]$TestName;
+    [ExecutableSpec]$Remote;
+    [ExecutableSpec]$Local;
+    [int]$Iterations;
+    [string]$RemoteReadyMatcher;
+    [string]$ResultsMatcher;
+
+    [string]ToString() {
+        $RetString = ("{0}_{1}_{2}_{3}" -f $this.TestName,
```

Comment:
`{${$this.TestName}}_{${$this.Remote.Platform}}_{${$script:RemoteTls}}_{${$script:RemoteArch}}`

What do you think about using this format instead? I'm not a big fan of the `-f`  method.